internal/api/promote_downstream_v1alpha1.go

```diff
@@ -160,16 +164,23 @@ func (s *server) PromoteDownstream(
 }
 
 // findDownstreamStages returns a list of Stages that are immediately downstream
-// from the given Stage.
+// from the given Stage and request Freight from the given origin.
 // TODO: this could be powered by an index.
-func (s *server) findDownstreamStages(ctx context.Context, stage *kargoapi.Stage) ([]kargoapi.Stage, error) {
+func (s *server) findDownstreamStages(
+	ctx context.Context,
+	stage *kargoapi.Stage,
+	origin kargoapi.FreightOrigin,
+) ([]kargoapi.Stage, error) {
 	var allStages kargoapi.StageList
 	if err := s.client.List(ctx, &allStages, client.InNamespace(stage.Namespace)); err != nil {
 		return nil, connect.NewError(connect.CodeInternal, err)
 	}
 	var downstreams []kargoapi.Stage
 	for _, s := range allStages.Items {
 		for _, req := range s.Spec.RequestedFreight {
+			if !req.Origin.Equals(&origin) {
+				continue
+			}
```

Comment:
Pretty sure it was an undiscovered bug that this check wasn't here already. I believe the potential has existed for Freight to be promoted to all of a Stage's downstreams _including_ those that may be uninterested in the Freight in question.